message_passing/waiting_for_handlers_and_compensation/workflows.py

```diff
@@ -0,0 +1,180 @@
+        self._workflow_compensation_done = False
+
+    @workflow.run
+    async def run(self, input: WorkflowInput) -> str:
+        try:
+            # 👉 Use this `try...except` style, instead of waiting for message
+            # handlers to finish in a `finally` block. The reason is that some
+            # exception types cause a workflow task failure as opposed to
+            # workflow exit, in which case we do *not* want to wait for message
+            # handlers to finish.
+
+            # 👉 self._run contains your actual application logic. This is
+            # implemented in a separate method in order to separate
+            # "platform-level" concerns (waiting for handlers to finish and
+            # ensuring that compensation is performed when appropriate) from
+            # application logic. In this sample, its actual implementation is
+            # below but contains nothing relevant.
+            result = await self._run(input)
+            self.workflow_exit.set_result(None)
+            await workflow.wait_condition(workflow.all_handlers_finished)
+            return result
+        # 👉 Catch BaseException since asyncio.CancelledError does not inherit
+        # from Exception.
+        except BaseException as e:
+            if is_workflow_exit_exception(e):
+                self.workflow_exit.set_exception(e)
+                await workflow.wait_condition(workflow.all_handlers_finished)
+                await self.workflow_compensation()
+                self._workflow_compensation_done = True
+            raise
+
+    async def workflow_compensation(self):
+        await workflow.execute_activity(
+            activity_executed_to_perform_workflow_compensation,
+            start_to_close_timeout=timedelta(seconds=10),
+        )
+        self._update_compensation_done = True
+
+    @workflow.update
+    async def my_update(self) -> str:
+        """
+        An update handler that handles exceptions raised in its own execution
+        and in that of the main workflow method.
+
+        It ensures that:
+        - Compensation/cleanup is always performed when appropriate
+        - The update caller gets the update result, or WorkflowUpdateFailedError
+        """
+        # 👉 As with the main workflow method, the update application logic is
+        # implemented in a separate method in order to separate "platform-level"
+        # error-handling and compensation concerns from application logic. Note
+        # that coroutines must be wrapped in tasks in order to use
+        # workflow.wait.
+        update_task = asyncio.create_task(self._my_update())
+
+        # 👉 "Race" the workflow_exit future against the handler's own application
+        # logic. Always use `workflow.wait` instead of `asyncio.wait` in
+        # Workflow code: asyncio's version is non-deterministic.
+        await workflow.wait(  # type: ignore
+            [update_task, self.workflow_exit], return_when=asyncio.FIRST_EXCEPTION
+        )
```

Comment:
This makes sense, and looks like a pain, but I'm not sure how often people want to cancel in-process update handlers in response to a workflow exit as opposed to simply waiting for the handlers to complete.
My guess is that most people are happy to wait because such cases are not particularly latency sensitive.
It may be okay that this is more of a power user struggle.

Reply:
I'm not sure I agree with this. The point isn't the latency -- it's that if the workflow is exiting due to a failure/cancellation then the update may need to perform a rollback / cancellation. So they can't "simply wait for the handler to complete" because the handler has to learn whether it should proceed to the end of its happy path, or perform rollback / compensation.